```{r config, echo=FALSE, message=FALSE}
modules::import('./scripts/knit', attach = TRUE)
opts_chunk$set(fig.path = 'figure/te-cu-vs-rcu-', cache = FALSE)
species = 'mouse'
celltype = 'E15.5'
```

# Measures of translation efficiency in `r species` `r celltype`

The following uses expression data from human liver. We will look at codon
usage of GO terms.

```{r load-data, echo=FALSE, message=FALSE, warning=FALSE}
config = import(paste0('./config_', species))
data = import('./data')

go_genes = data$go_genes(config)
canonical_cds = data$canonical_cds(config)

mrna_library = data$mrna_design(config) %>%
    filter(Celltype == celltype) %>% .$DO %>% first()
mrna_counts = data$mrna_counts(config) %>%
    select_(~Gene, ~Name, Count = mrna_library) %>%
    inner_join(canonical_cds, by = 'Gene')

trna_library = data$trna_design(config) %>%
    filter(Celltype == celltype) %>% .$DO %>% first()
trna_counts = data$trna_counts(config) %>%
    select_(~Gene, AminoAcid = ~AA, ~Anticodon, Count = trna_library)
```

## Total codon usage (CU)

“Translation efficiency” (TE) measures how well codon usage and anticodon
abundance are adapted. Specifically, whether a given codon can be efficiently
translated because its concomitant tRNA anticodon is efficiently found by the
ribosome. This requires the tRNA to be abundant in just the right ratio — that
dictated by codon usage, relative to the usage of other codons.

We therefore scale codon usage (CU) and anticodon abundance (AA) by the total
codon usage and anticodon abundance. The TE is then defined as the simple (rank)
correlation of scaled CU and AA.

```{r cu}
cu_ = import('codon_usage')
cu = mrna_counts %>%
    cu_$cu() %>%
    inner_join(go_genes, by = 'Gene') %>%
    group_by(GO, Codon) %>%
    summarize(CU = sum(CU)) %>%
    mutate(CU = CU / sum(CU))

aa = trna_counts %>%
    group_by(Anticodon) %>%
    summarize(AA = sum(Count)) %>%
    mutate(AA = AA / sum(AA)) %>%
    mutate(Codon = cu_$revcomp(Anticodon))

te_cu = inner_join(cu, aa, by = 'Codon') %>%
    summarize(TE = cor(CU, AA, method = 'spearman'))
```

We now calculate calculate the genomic background CU and correlate it with the
GO term CUs (this corresponds to the “size factor effect” plot).

```{r background-cu}
background_cu = canonical_cds %>%
    cu_$cu() %>%
    group_by(Codon) %>%
    summarize(CU = sum(CU)) %>%
    mutate(CU = CU / sum(CU))

background_cor = inner_join(cu, background_cu, by = 'Codon') %>%
    group_by(GO) %>%
    summarize(Correlation = cor(CU.x, CU.y, method = 'spearman')) %>%
    inner_join(group_by(go_genes, GO) %>% summarize(Size = n()), by = 'GO')

ggplot(background_cor, aes(x = Size, y = Correlation)) +
    geom_point() +
    scale_x_log10()
```


## Relative codon usage (RCU)

A different approach to calculating the TE is to look at the selective
pressure exerted on the codon deployment: that is, how abundant a given codon
is compared to synonymous codons (ignoring wobble base pairing). This compels
us to look at relative codon usage (RCU) within a synonymous codon family and,
conversely, at the relative anticodon abundance (RAA) within a tRNA anticodon
isoacceptor family.

```{r rcu}
rcu = mrna_counts %>%
    cu_$cu() %>%
    inner_join(cu_$genetic_code, by = 'Codon') %>%
    inner_join(go_genes, by = 'Gene') %>%
    group_by(GO, AA, Codon) %>%
    summarize(CU = sum(CU)) %>%
    mutate(RCU = CU / sum(CU)) %>%
    group_by(GO)

raa = trna_counts %>%
    group_by(AminoAcid, Anticodon) %>%
    summarize(RAA = sum(Count)) %>%
    mutate(RAA = RAA / sum(RAA)) %>%
    ungroup() %>%
    mutate(Codon = cu_$revcomp(Anticodon))

te_rcu = inner_join(rcu, raa, by = 'Codon') %>%
    summarize(TE = cor(RCU, RAA, method = 'spearman'))
```

Let’s plot the distributions.

```{r boxplot-te}
te = bind_rows(mutate(te_cu, Mode = 'CU vs AA'),
               mutate(te_rcu, Mode = 'RCU vs RAA'))

ggplot(te) + geom_boxplot(aes(Mode, TE))
```

We then calculate the correlation between CU-TE and RCU-TE:

```{r correlate-te}
te_cor = inner_join(te_cu, te_rcu, by = 'GO') %>%
    rename(`CU vs AA` = TE.x, `RCU vs RAA` = TE.y)

ggplot(te_cor, aes(x = `CU vs AA`, y = `RCU vs RAA`)) +
    geom_point() +
    geom_smooth(method = lm)

cor(te_cu$TE, te_rcu$TE, method = 'spearman')
```
